### ISSUE TITLE
Rename `/sections` route to `/sections/transform`

### ISSUE THREAD

**Issue by jamesscottbrown** [ove/ove#85]
OVE core currently accepts `POST`, `GET` and `DELETE` HTTP methods on `/sections`.

 I'm happy with how the `GET` and `DELETE` actions are handled, but think that the `POST` action is problematic.

`GET` returns details of the current state of sections. The example response body is

```JSON
[
  {
    "id": 0,
    "x": 0,
    "y": 0,
    "w": 30720,
    "h": 4320,
    "space": "LocalNine",
    "app": {
      "url": "http://192.168.1.108:8084"
    }
  }
]
```

One would naively expect the `POST` action to accept updated objects with the same format, but it actually requires details of a *transform operation*, such as:

```JSON
{
  "scale": {
    "x": 0,
    "y": 0
  },
  "translate": {
    "x": 0,
    "y": 0
  }
}
```

This is not simply a POST request to create or replace a resource, but rather a request to modify an existing resource in a particular way, and the name of the API route should reflect this. The current name suggests that I could `POST` to `/sections` to create multiple sections as an alternative to repeatedly `POST`ing to `/section`, which is wrong.

I suggest something like `/sections/transform`.

**Comment by senakafdo** [comment]
You have a point, but there are two sides to this.

So, in REST you modify a resource, and in this case, the resource is '/sections'. POST defines the operation you would like to perform on the resource whereas PUT represents the content you would like place at the resource. Our API is therefore valid [according to the HTTP specification](https://tools.ietf.org/html/rfc7231#section-4.3.3) and should not confuse a user.

However, we could also provide a separate method to create/update all sections in one go, using a [PUT method](https://tools.ietf.org/html/rfc7231#section-4.3.4). But, I feel that we do not need to be doing that.

**Comment by senakafdo** [comment]
> As the spec you linked to states, a PUT method "requests that the enclosed entity be stored *under the supplied Request-URI*". However, the URI at which a section is created is chosen by the sever (which increments an id counter), and the client has no control over this. Creating a set of sections at once should thus use a POST request (as we are currently using to create sections and groups).

If you read carefully, they note the following:

<code>
The PUT method requests that the state of the target resource be <strong>created</strong> or <strong>replaced</strong> with the state defined by the representation enclosed in the request message payload.
</code>
<br/><br/>
<code>
If the target resource <strong>does have a current representation</strong> and that representation is successfully modified in accordance with the state of the enclosed representation, then the origin server MUST send either a 200 (OK) or a 204 (No Content) response to indicate successful completion of the request.
</code>
<br/><br/>
<code>
The fundamental difference between the POST and PUT methods is highlighted by the different intent for the enclosed representation. The target resource in a POST request is intended to handle the enclosed representation according to the resource's own semantics, whereas the enclosed representation in a PUT request is defined as replacing the state of the target resource.  Hence, the intent of PUT is idempotent and visible to intermediaries, even though the exact effect is only known by the origin server.
</code>
<br/><br/>

> ideally we would adopt the Principle of Least Astonishment

Well, yes, but POLA is not so relevant in this instance. As explained in this [blog post](http://www.monitis.com/blog/apis-and-the-principle-of-least-surprise/), it concerns bad practices in API design such as throwing `not implemented` exceptions.

> There's also a larger API design issue here: the transformation applies to a group or a space rather than acting on the whole collection of resources

I believe you have not fully understood this operation or carefully read the documentation. The documentation again is clear on _`This operation will be limited to a subset of sections if a group id or a space was provided`_. So, if you did not provide either, it will actually apply to the whole collection of resources.

I did however realise that the operation represents **move to a new space** as a transformation, which is slightly confusing. Therefore I clearly separated them out in 31b7a5c.

> I'm happy to shelve this for now

Yes. Beyond 31b7a5c, I don't find it productive to be debating on this. And, there are a number of other things that we can be fixing in the meantime.

**Comment by jamesscottbrown** [comment]
After our meeting today, we have decided to add new routes:

* `/spaces/<id>/transform`
* `/groups/<id>/transform`
* `/sections/<id>/transform`

We will also change `POST /sections` to create a list of sections (analagously to how `GET` and `DELETE` on `/sections` is equivalent to separately acting on the corresponding `/section`s)

**Comment by senakafdo** [comment]
@jamesscottbrown, further to the last chat we had, lets first review the entire API documentation that we have and then decide on what's the best approach/naming. I have a feeling that even the above may not fit well with the rest of our APIs, since all other operations on sections work on a model where you provide space/group details in the query. So, instead of fixing this as a one-off, lets have a read at the entire API and decide what's best.

**Comment by senakafdo** [comment]
@jamesscottbrown, sorry for repeatedly disagreeing to fix anything here, but your request once again falls short of a few important things and puts in a much worse situation that we are at with the current API.
1. There are a number of operations you can perform on a set of sections that would result in them being updated and transform is just one of them. There is also a moveTo operation, and if we keep adding API methods for all of these it is going be a challenge down the line, especially when maintaining backwards compatibility etc - so fewer the better along these lines.
2. You could actually perform a transform or whatever other update operation on all sections (such as move them all to a different space regardless of what spaces they may be in right now), which is possible with the current design but not possible with the one that has been proposed. So, in the process, we have lost some functionality and that is also not good.
3. Also, when it comes to the URLs, `/spaces/<id>/transform` and `/groups/<id>/transform` the operation does not make sense being at that level, since you are actually transforming `sections in a space` or `sections in a group` and not the `space` or `group`. So, ideally, we should then have, `/spaces/<id>/sections/transform` which again is quite similar to `/sections/transform`. This again, ends having the same issues in terms of not following best practices in API design etc. There is however the possibility, of implementing more than one type of transformation operation on a space and group, but that would make their payloads very different to each which may lead into other issues.

Also, at the same time, your proposal for aggregate updates on `sections` with a POST will also end up having confusing semantics if we go ahead and do that. Consider the following scenarios.
1. I have ten sections already and I am making a POST with one new section, do we append it and have eleven sections, or do we delete the existing ten and have one instead
2. I have four sections and I have done nothing much re-ordering them in my update. Now, do we delete and recreate all sections, or do we intelligently make changes to sections. What would the end user be expecting under these circumstances, and what would be practically possible given that section ids are stored in states of all apps, which we will anyway have to clean up and restore.
3. I have three sections, and I'm submitting a POST request with two of those three sections. Isn't this another form of DELETE that I have just managed to achieve but working in a very different way.

I would therefore like to point out that a POST is as easy as DELETE or GET when it comes to avoiding amguity and the way things work today at least is consistent though not necessarily convenient.

I however agree that our API design may not be the most optimum. It has never been reviewed in detail and we also do not have clarity as to whether we have understood its entire scope to say if X is correct and Y is not in terms of alternatives.

Therefore, what I would rather do is close this issue right now, which anyway is not wrong since what you have proposed (even for the second time) does not work. And, at the same time, I will create a separate issue to ensure we thoroughly review the entire API not just for OVE Core but also across the board and standardise that, as a part of the 1.0.0 release, the day we are ready for that. I hope you'd be fine with this approach.

**Comment by senakafdo** [comment]
Closing issue as this won't be fixed as requested, and since a new issue (#96) has been reported to tackle the problem across the board.
